Docs updates for Plotly.py version 6 (#4840)
plotly/plotly.py: doc/python/px-arguments.md

    thumbnail: thumbnail/plotly-express.png
---

Plotly Express (`px`) is the high-level interface to Plotly and provides functions for generating charts. `px` functions support data provided in a number of different formats (long, wide, and mixed) and as different types of objects, including pandas and Polars DataFrames.

## Data for the Examples

The examples on this page use datasets available in the `data` package in `px`. `px.data` contains functions that when called return a dataset as a DataFrame. Some of the datasets included in `px.data` are:

- `carshare` - Each row represents the availability of car-sharing services near the centroid of a zone in Montreal over a month-long period.
- `election` - Each row represents voting results for an electoral district in the 2013 Montreal mayoral election.
- `iris` - Each row represents a flower.

> To access the `iris` dataset, we call it's function and assign it to a variable:

Reviewer: "its" not "it's"

```python
import plotly.express as px

df = px.data.iris()
df.head()
```

> By default `px.data` functions return Pandas DataFrames, but you can specify an alternative DataFrame type using `return_type`. `pandas`, `polars`, and `pyarrow` are supported return types.

Reviewer: "dataframe type" not "DataFrame type"

```python
df = px.data.iris(return_type='polars')
df.head()
```

> ## Long, Wide-, and Mixed-Form Data

Reviewer: puzzled: why no dash after Long when there is one after Wide? (I wouldn't have a dash after either.)

### Supported DataFrame Types

`px` functions natively support pandas, Polars, and PyArrow dataframes. `px` uses [Narwhals](https://narwhals-dev.github.io/narwhals/) to provide this native dataframe support. Other types of dataframes that are currently supported by Narwhals, for example cuDF and Modin, may also work with `px`.

You can also pass dataframes that are not natively supported, but which support the [dataframe interchange protocol](https://data-apis.org/dataframe-protocol/latest/).

PySpark dataframes are also supported and are converted to pandas dataframes internally by Plotly Express.

#### Additional Dependencies Required

- Plotly Express requires NumPy. You can install it with `pip install numpy` if it's not installed by the dataframe library you are using.
- To use [trendlines](/python/linear-fits/), you'll also need to have pandas installed.
- To use PySpark dataframes, you'll need to have pandas installed. To use dataframes that support the dataframe interchange protocol, you'll need to have PyArrow installed.

> ### Example: Using a Pandas DataFrame with `px.bar`

Author: ```suggestion

### Example: Using a Pandas DataFrame with `px.bar`
```

Many more examples of wide-form and messy data input can be found in our [detailed wide-form support documentation](/python/wide-form/).

## Dataframe Input

The first argument of every `px` function is `data_frame`. If you provide a dataframe as a `px` function's first argument, you can then specify column names as strings from the dataframe as other arguments.

> ### Supported Dataframes

Reviewer 2: ```suggestion
### Supported DataFrame Types
```

`px` functions natively support the following types of DataFrames:

> - pandas

Reviewer: "Pandas" rather than "pandas" (?) not sure what the capitalization should be

Author: I say "pandas" as that's how it is in their docs: https://pandas.pydata.org/docs/user_guide/index.html

> `px` functions natively support pandas, Polars, and PyArrow dataframes. `px` uses [Narwhals](https://narwhals-dev.github.io/narwhals/) to provide this native dataframe support. Other types of dataframes that are currently supported by Narwhals, for example cuDF and Modin, may also work with `px`.

Reviewer 2: Note: Need to make sure we are consistent with messaging about whether Modin is supported (it's included in the list of supported DataFrame types in `doc/python/performance.md`)

Author: Good catch. I'd changed it in one but not the other. Thanks!

- Polars
- PyArrow
- Modin

> To use Polars, PyArrow, or Modin with Plotly Express, you'll need to have NumPy installed. You can install it with: `pip install numpy`.
To use [trendlines](/python/linear-fits/), you'll also need to have pandas installed. 

`px` uses [Narwhals](https://narwhals-dev.github.io/narwhals/) to natively support these DataFrames. Other types of DataFrames that are currently supported by Narwhals may also work with `px`, for example cuDF.

> You can also pass DataFrames that are not natively supported, but support the dataframe interchange protocol, or have a `to_pandas()` method.

Reviewer: again, do we use "DataFrame" for the generic object or "dataframe"?

> - To use Polars, PyArrow, or other dataframes supported natively with Plotly Express, you'll need to have NumPy installed. You can install it with: `pip install numpy`.

Reviewer 2: A little misleading, since NumPy is required for Pandas dataframes as well, it's just that NumPy is a dependency of Pandas so it's likely to be already installed if you're instantiating a Pandas dataframe.

Author: That's a fair point. Updated

> - To use PySpark dataframes or dataframes that support the dataframe interchange protocol, you'll need to have pandas installed.
> 

Author: ```suggestion
- To use PySpark dataframes, you'll need to have pandas installed. To use dataframes that support the dataframe interchange protocol, you'll need to have PyArrow installed.
```

Many more examples of wide-form and messy data input can be found in our [detailed wide-form support documentation](/python/wide-form/).

## Dataframe Input

The first argument of every `px` function is `data_frame`. If you provide a dataframe as a `px` function's first argument, you can then specify column names as strings from the dataframe as other arguments.

### Supported Dataframes

`px` functions natively support the following types of dataframes:

- pandas
- Polars
- PyArrow
- Modin

> `px` uses [Narwhals](https://narwhals-dev.github.io/narwhals/) to natively support these dataframes. Other types of dataframes that are currently supported by Narwhals may also work with `px`.

Reviewer: Move this up above the list to replace the "px functions natively support..." sentence? That feels redundant given this more detailed explanation.

Author: Reordered the content

fig.show()
```

### Using Columns from Multiple DataFrames

> You can also use columns from multiple dataframes in the one `px` function, as long as all the DataFrame columns you use have the same length. In this example, we pass `df1` as the `data_frame` argument to `px.bar` and then us a column from `df2` for the `y` argument.

Reviewer: "in one" as opposed to "in the one"

### Passing dictionaries or array-likes as the data_frame argument

> The column-based argument `data_frame` can also be passed with a `dict` or `array`. Using a dictionary can be a convenient way to pass column names used in axis titles, legend entries and hovers without creating a DataFrame.

Reviewer: why "data_frame" in this case?

Author: In this case it's name of the px parameter.

> The column-based argument `data_frame` can also be passed with a `dict` or `array`. Using a dictionary can be a convenient way to pass column names used in axis titles, legend entries and hovers without creating a dataframe.

Reviewer 2: ```suggestion
The `data_frame` argument can also accept a `dict` or `array` in addition to DataFrame objects. Using a dictionary can be a convenient way to pass column names used in axis titles, legend entries and hovers without creating a dataframe.
```
